parkings/api/enforcement/active_parking.py

```diff
@@ -0,0 +1,111 @@
+
+    class Meta:
+        model = Parking
+        fields = [
+            'id',
+            'created_at',
+            'modified_at',
+            'registration_number',
+            'time_start',
+            'time_end',
+            'zone',
+            'operator_name',
+        ]
+
+
+class ActiveParkingFilter(django_filters.rest_framework.FilterSet):
+    reg_num = django_filters.CharFilter(
+        name='reg_num', label=_("Registration number"),
+        method='filter_reg_num', required=True)
+    time = django_filters.IsoDateTimeFilter(
+        name='time', label=_("Time"),
+        method='filter_time')
+
+    class Meta:
+        model = Parking
+        fields = []
+        strict = django_filters.STRICTNESS.RAISE_VALIDATION_ERROR
+
+    def __init__(self, data=None, *args, **kwargs):
+        """
+        Initialize filter set with given query data.
+
+        :type data: django.http.request.QueryDict
+        """
+        if data and 'time' not in data:
+            data = data.copy()
+            data.setdefault('time', timezone.now())
+        super().__init__(data, *args, **kwargs)
+
+    def filter_reg_num(self, queryset, name, value):
+        """
+        Filter by normalized registration number.
+
+        :type queryset: django.db.models.QuerySet
+        :type name: str
+        :type value: str
+        """
+        reg_num = Parking.normalize_reg_num(value)
+        return queryset.filter(normalized_reg_num=reg_num)
+
+    def filter_time(self, queryset, name, value):
+        """
+        Filter to active parkings at given time stamp.
+
+        If there is no active parkings at given time, but there is a
+        parking within a day from given time, then return the parking
+        that has the latest ending time.
+
+        :type queryset: django.db.models.QuerySet
+        :type name: str
+        :type value: datetime.datetime
+        """
+        time = value if value else timezone.now()
+        active_parkings = _filter_time_range(
+            queryset, starts_before=time, ends_after=time)
+        if not active_parkings:
+            active_within_limit = _filter_time_range(
+                queryset,
+                starts_before=time,
+                ends_after=(time - datetime.timedelta(days=1)))
+            return active_within_limit.order_by('-time_end')[:1]
+        return active_parkings
+
+
+def _filter_time_range(queryset, starts_before, ends_after):
+    """
+    Filter queryset by its start/end time values.
+
+    Filter to objects that start before the starts_before value and end
+    after the ends_after value or have no ending time set.
+
+    :type queryset: django.db.models.QuerySet
+    :type starts_before: datetime.datetime
+    :type ends_after: datetime.datetime
+    :rtype: django.db.models.QuerySet
+    """
+    return queryset.filter(
+        Q(time_start__lte=starts_before) &
+        Q(time_end__gte=ends_after) | Q(time_end__isnull=True))
```

Comment:
The `OR` would be nice to be inside parenthesis for extra clarity and to avoid future accidental problems.

Reply:
Oops, good catch. It was meant to be so.